tinygrad/shape/symbolic.py

```diff
@@ -191,11 +229,36 @@ def vars(self): return functools.reduce(lambda l,x: l+x.vars(), self.nodes, [])
 
 class SumNode(RedNode):
   def __mul__(self, b: Union[Node, int]): return Node.sum([x*b for x in self.nodes]) # distribute mul into sum
-  def __floordiv__(self, b: int, factoring_allowed=True):
-    if b == 1: return self
-    if not factoring_allowed: return Node.__floordiv__(self, b, factoring_allowed)
+  def __floordiv__(self, b: Union[Node, int], factoring_allowed=True):
     fully_divided: List[Node] = []
     rest: List[Node] = []
+    if isinstance(b, SumNode):
+      numerator_num = sum([node.b for node in self.flat_components if node.__class__ is NumNode])
+      denominator_num = sum([node.b for node in b.flat_components if node.__class__ is NumNode])
+      if numerator_num and denominator_num and numerator_num % denominator_num == 0:
+        attempt = numerator_num // denominator_num
+        if b * attempt - self == 0: return NumNode(attempt)
+    elif isinstance(b, MulNode) and isinstance(b.b, int):  # try to divide b.b if it's int
+      fully_divided, rest = [], []
+      for x in self.flat_components:
+        if x % b.b == 0: fully_divided.append(x // b.b)
+        else: rest.append(x)
+      rest_sum = create_rednode(SumNode, rest)
+      if b - rest_sum.max > 0: return create_rednode(SumNode, fully_divided) // b.a
+    if isinstance(b, Node):  # try to divide the whole b
+      fully_divided, rest = [], []
+      for x in self.flat_components:
+        try:
+          if x % b == 0: fully_divided.append(x // b)
+          else: rest.append(x)
+        except RuntimeError:
+          rest.append(x)
+      rest_sum = create_rednode(SumNode, rest)
+      if b - rest_sum.max > 0: return create_rednode(SumNode, fully_divided)
```

Comment:
```python
x = Variable("x", 0, 10)
z = Variable("z", 100, 200)

div = (x - 1000) // z # should not be 0
mod = (x - 1000)  % z # should not be `self` or `(x - 1000)`
```

You missed lower bound checks (`b < rest_sum.min`).

You can get away with these checks in `Node` (because negative `self.min` is not allowed) and `MulNode` (because it is symmetric).

Reply:
*`-b < rest_sum.min` or `b > -rest_sum.min`

Reply:
(not reproducible now but was in 42c739b and 0fe885e)

oh wait, i got fooled by `MulNode`. It's not worth it to omit these checks anywhere, just check everywhere.

```python
(x*-100) % z # should not be (x*-100)
```

Reply:
Also note that you can't have `LtNode`s in an `and` expression, only concrete deterministic `bool`s will give correct answers.